pkg/upstream/cluster/cluster_manager.go

```diff
@@ -62,12 +62,80 @@ const globalTLSMetrics = "global"
 // types.ClusterManager
 type clusterManager struct {
 	clustersMap      sync.Map
-	protocolConnPool sync.Map // protocolname: { address : connpool }
+	protocolConnPool *connPool
 	tlsMetrics       *mtls.TLSStats
 	tlsMng           atomic.Value // store types.TLSClientContextManager
 	mux              sync.Mutex
 }
 
+type connPool struct {
+	clusterPoolEnable bool     // TODO support modify clusterPoolEnable at runtime
+	globalPool        sync.Map // proto: {addr: pool}
+	clusterPool       sync.Map // proto: {cluster: {addr: pool}}
+}
+
+func newConnPool(clusterPoolEnable bool) *connPool {
+	return &connPool{
+		clusterPoolEnable: clusterPoolEnable,
+	}
+}
+
+func (p *connPool) store(protocolName api.ProtocolName) {
+	p.globalPool.Store(protocolName, &sync.Map{})
+	p.clusterPool.Store(protocolName, &sync.Map{})
+}
+
+func (p *connPool) load(proto types.ProtocolName, snapshot types.ClusterSnapshot) (*sync.Map, bool) {
+	var connectionPool *sync.Map
+	if p.clusterPoolEnable || snapshot.ClusterInfo().IsClusterPoolEnable() {
+		if poolMap, ok := p.clusterPool.Load(proto); ok {
+			value, _ := poolMap.(*sync.Map).LoadOrStore(snapshot.ClusterInfo().Name(), &sync.Map{})
+			connectionPool = value.(*sync.Map)
+			return connectionPool, ok
+		}
+	} else {
+		if poolMap, ok := p.globalPool.Load(proto); ok {
+			connectionPool = poolMap.(*sync.Map)
+			return connectionPool, ok
+		}
+	}
+	return nil, false
+}
+
+// getConnPool return clusterProtoPool and globalProtoPool if proto is valid
+func (p *connPool) getConnPool(proto types.ProtocolName) (clusterProtoPool interface{}, globalProtoPool interface{}, ok bool) {
+	if clusterProtoPool, ok = p.clusterPool.Load(proto); !ok {
+		return
+	}
+	if globalProtoPool, ok = p.globalPool.Load(proto); !ok {
+		return
+	}
+	return
+}
+
+func (p *connPool) shutdown(proto types.ProtocolName, addr string) {
+	if proto == "" {
+		p.clusterPool.Range(func(key, value interface{}) bool {
+			shutdownClusterPool(value, addr, key.(api.ProtocolName))
```

Comment:
Let's see if there's a better way. maybe `shutdownClusterPool()` and `shutdownGlobalPool()` should belong to `connPool`.

Reply:
how about merge `shutdownClusterPool()` and `shutdownGlobalPool()` into `connPool#shutdownPool`? Such as:
```go
func (p *connPool) shutdownAll(proto types.ProtocolName, addr string) {
	shutdownPool := func(value interface{}) {
		connectionPool := value.(*sync.Map)
		if cp, ok := connectionPool.Load(addr); ok {
			pool := cp.(types.ConnectionPool)
			connectionPool.Delete(addr)
			pool.Shutdown()
		}
	}
	clusterProtoPool, globalProtoPool, _ := p.getConnPool(proto)
	clusterProtoPool.(*sync.Map).Range(func(_, connPool interface{}) bool {
		shutdownPool(connPool)
		return true
	})
	shutdownPool(globalProtoPool)
}
```

Reply:
good idea ~

Reply:
` shutdownAll` rename to `shutdown`. lgtm~